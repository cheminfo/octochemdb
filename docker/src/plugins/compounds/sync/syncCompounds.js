import { main } from '../fix/main.js';

import firstCompoundsImport from './firstCompoundsImport.js';
import incrementalCompoundsImport from './incrementalCompoundsImport.js';
/**
 * @description Synchronize the compounds database from the pubchem database
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns compounds collections
 */
export async function sync(connection) {
  main(connection);
  // Do the first importation
  //await firstCompoundsImport(connection);
  // Do the incremental updates
  // await incrementalCompoundsImport(connection);
}
