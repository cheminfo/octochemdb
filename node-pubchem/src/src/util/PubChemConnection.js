'use strict';

const delay = require('delay');
const { MongoClient } = require('mongodb');
const debug = require('debug')('PubChemConnection');

function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function getProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.updateOne({ _id: progress._id }, { $set: progress });
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    debug('Starting new database construction.');
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  } else {
    if (progress.state === 'update') {
      debug('First importation has been completed. Should only update.');
    } else {
      debug(`Continuing first importation from ${progress.seq}.`);
    }
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};

module.exports = PubChemConnection;
