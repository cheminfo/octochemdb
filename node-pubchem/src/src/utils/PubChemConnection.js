import Debug from 'debug';
import delay from 'delay';
import { MongoClient } from 'mongodb';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
