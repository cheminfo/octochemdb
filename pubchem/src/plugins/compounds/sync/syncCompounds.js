import firstCompoundsImport from './firstCompoundsImport.js';
import incrementalCompoundsImport from './incrementalCompoundsImport.js';

export async function sync(connection) {
  await firstCompoundsImport(connection);
  await incrementalCompoundsImport(connection);
}
