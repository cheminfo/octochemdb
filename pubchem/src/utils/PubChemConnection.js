import delay from 'delay';
import { MongoClient } from 'mongodb';

import Debug from './Debug.js';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getLogs = async function getCollection(
  collectionName,
) {
  const logsCollection = await this.getLogsCollection();
  const _id = `${collectionName}`;
  let logs = await logsCollection.find({ _id }).next();
  if (logs === null) {
    let sources = [];
    logs = {
      _id,
      sources,
      sourcesHash: md5(JSON.stringify(sources)),
      dataStart: 0,
      dataEnd: 0,
      startSequenceID: 123,
      endSeqneceID: 123,
      status: 'updating',
    };
    await logsCollection.insertOne(logs);
  }
  return logs;
};
PubChemConnection.prototype.setLogs = async function setLogs(logs) {
  const collection = await this.getLogsCollection();
  await collection.replaceOne({ _id: logs._id }, logs);
};
PubChemConnection.prototype.getLogsCollection = async function getCollection() {
  return (await this.getDatabase()).collection('logs');
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
