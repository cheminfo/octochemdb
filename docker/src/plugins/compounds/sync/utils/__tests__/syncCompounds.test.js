import { PubChemConnection } from '../../../../../utils/PubChemConnection.js';
import { sync } from '../../syncCompounds';

test('syncCompounds First Importation', async () => {
  const connection = new PubChemConnection();
  await sync(connection);
  const collection = await connection.getCollection('compounds');
  const collectionEntry = await collection.find({ _id: 160056959 }).limit(1);
  const result = await collectionEntry.next();
  expect(result).toMatchSnapshot();
  if (connection) {
    await connection.close();
  }
});
test('syncCompounds Incremental Importation', async () => {
  const connection = new PubChemConnection();
  const collection = await connection.getCollection('compounds');
  const collectionEntryIncremental = await collection
    .find({ _id: 160056959 })
    .limit(1);
  const resultIncremental = await collectionEntryIncremental.next();
  expect(resultIncremental).toMatchSnapshot();
  if (connection) {
    await connection.close();
  }
});
