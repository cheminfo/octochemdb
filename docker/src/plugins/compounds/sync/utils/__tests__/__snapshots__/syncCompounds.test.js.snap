// Vitest Snapshot v1

exports[`syncCompounds First Importation 1`] = `
{
  "_id": 59478,
  "data": {
    "atom": "{\\"C\\":37,\\"H\\":60,\\"O\\":8}",
    "charge": 0,
    "em": 632.42881889036,
    "inchi": "InChI=1S/C12H9N3S2/c13-8-1-3-9(4-2-8)16-10-5-6-11-12(7-10)15-17-14-11/h1-7H,13H2",
    "inchiKey": "OPAAXBCEKMMTKB-UHFFFAOYSA-N",
    "iupac": "4-(2,1,3-benzothiadiazol-5-ylsulfanyl)aniline",
    "mf": "C37H60O8",
    "molfile": "59478
  -OEChem-01162303312D

45 50  0  0  0  0  0  0  0  0999 V2000
   11.6583   -3.8241   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   10.8878   -3.1866   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   11.0547   -2.2006   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   10.2842   -1.5632   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   10.4511   -0.5772   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.6806    0.0603   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.6388    1.0594   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    8.6757    1.3284   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.1222    0.4956   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.7433   -0.2882   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.5938   -1.1634   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.3469   -1.9117   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.3969    1.0657   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.4187    0.8578   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.1097   -0.0933   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.7788   -0.8364   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7008   -0.8015   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.3686   -1.5576   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.3786   -1.2577   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.2788   -1.7024   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.3007   -1.4945   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.1962   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.3301   -0.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4641   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981   -0.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981    1.0000   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321    1.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8660   -0.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8660    1.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000   -0.5000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.3301    1.0000   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.2951   -1.6928   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.3666    0.5759   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   11.4370   -0.7440   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   11.0885    0.1933   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.9505   -3.5351   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.1801   -2.8976   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    9.7837   -4.5211   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.8464   -4.8696   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   10.5542   -5.1586   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   11.4915   -4.8101   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   12.2619   -5.4476   -0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   13.1992   -5.0991   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   10.3873   -6.1446   -0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
  2  1  1  0  0  0  0
  3  2  1  0  0  0  0
  4  3  1  0  0  0  0
  5  4  1  0  0  0  0
  6  5  1  0  0  0  0
  7  6  1  0  0  0  0
  8  7  1  0  0  0  0
  9  8  1  0  0  0  0
 10  9  1  0  0  0  0
 10  6  1  0  0  0  0
 11 10  1  0  0  0  0
 12 11  1  0  0  0  0
 12  4  1  0  0  0  0
 13  9  1  0  0  0  0
 14 13  1  0  0  0  0
 15 14  1  0  0  0  0
 16 15  1  0  0  0  0
 17 16  1  0  0  0  0
 17  9  1  0  0  0  0
 18 17  1  0  0  0  0
 19 18  2  0  0  0  0
 19  6  1  0  0  0  0
 20 16  1  0  0  0  0
 21 20  1  0  0  0  0
 22 21  1  0  0  0  0
 22 15  1  0  0  0  0
 23 22  1  0  0  0  0
 24 23  1  0  0  0  0
 25 24  1  0  0  0  0
 26 25  1  0  0  0  0
 27 26  1  0  0  0  0
 28 25  1  0  0  0  0
 29 28  2  0  0  0  0
 30 29  1  0  0  0  0
 31 29  1  0  0  0  0
 32 23  1  0  0  0  0
 33 16  1  0  0  0  0
 34 15  1  0  0  0  0
 35  5  1  0  0  0  0
 36  5  1  0  0  0  0
 37  2  1  0  0  0  0
 38 37  1  0  0  0  0
 39 37  1  0  0  0  0
 40 39  1  0  0  0  0
 41 39  1  0  0  0  0
 42 41  1  0  0  0  0
 42  1  1  0  0  0  0
 43 42  1  0  0  0  0
 44 43  1  0  0  0  0
 45 41  1  0  0  0  0
M  END
",
    "mw": 632.868912817097,
    "nbFragments": 1,
    "ocl": {
      "acceptorCount": 8,
      "coordinates": "!BkDKv^?DDiHVdaYL}kDJlPoY{}gn[noDDiHVdaZqB^?D}kFslwoaF_zREQJYxN\\\\KL?Fr\\\\DfQAaLg|{OIKcTodY^s|aMXt|cSu?o\`t|o{xMOTt|cSrMOW~~@",
      "donorCount": 4,
      "idCode": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
      "index": [
        229,
        79,
        253,
        255,
        254,
        94,
        170,
        237,
        80,
        175,
        254,
        179,
        118,
        206,
        189,
        138,
        45,
        251,
        196,
        249,
        52,
        196,
        107,
        122,
        203,
        93,
        128,
        71,
        198,
        146,
        82,
        212,
        156,
        2,
        100,
        120,
        1,
        151,
        2,
        165,
        4,
        88,
        0,
        18,
        131,
        5,
        3,
        128,
        1,
        16,
        8,
        8,
        64,
        81,
        112,
        48,
        64,
        0,
        0,
        32,
        4,
        128,
        0,
        8,
      ],
      "logP": 4.544999875128269,
      "logS": -5.477999992668629,
      "noStereoID": "ekTpA@@BKBiod\\\\Q\\\\ddTbTRaURTTfbTVfQratRTdtd\\\\TLRjV^vnjjfjjjjjjjjjjfjjh@@L@",
      "noStereoTautomerID": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
      "polarSurfaceArea": 117.83999633789062,
      "rotatableBondCount": 8,
      "stereoCenterCount": 15,
    },
    "unsaturation": 8,
  },
}
`;

exports[`syncCompounds Incremental Importation 1`] = `
{
  "_id": 160056959,
  "data": {
    "atom": "{\\"C\\":12,\\"H\\":9,\\"N\\":3,\\"S\\":2}",
    "charge": 0,
    "em": 259.02378965215996,
    "inchi": "InChI=1S/C12H9N3S2/c13-8-1-3-9(4-2-8)16-10-5-6-11-12(7-10)15-17-14-11/h1-7H,13H2",
    "inchiKey": "OPAAXBCEKMMTKB-UHFFFAOYSA-N",
    "iupac": "4-(2,1,3-benzothiadiazol-5-ylsulfanyl)aniline",
    "mf": "C12H9N3S2",
    "molfile": "166451518
  -OEChem-01312300422D

 26 28  0     0  0  0  0  0  0999 V2000
    6.0010    1.3100    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
   10.1289    0.3100    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    9.5453    1.1147    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    9.5453   -0.4947    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    2.5369   -0.6900    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    6.8671    0.8100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.5991    0.8100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.5991   -0.1900    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7331    1.3100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.1350    0.8100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.8671   -0.1900    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7331   -0.6900    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.1350   -0.1900    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.2690    1.3100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4030   -0.1900    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.2690   -0.6900    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4030    0.8100    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7331    1.9300    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    6.3301   -0.5000    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    7.7331   -1.3100    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    5.6720   -0.5000    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.2690    1.9300    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    4.2690   -1.3100    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.8660    1.1200    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.0000   -0.3800    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.5369   -1.3100    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  6  1  0  0  0  0
  1 10  1  0  0  0  0
  2  3  1  0  0  0  0
  2  4  1  0  0  0  0
  3  7  2  0  0  0  0
  4  8  2  0  0  0  0
  5 15  1  0  0  0  0
  5 25  1  0  0  0  0
  5 26  1  0  0  0  0
  6  9  2  0  0  0  0
  6 11  1  0  0  0  0
  7  8  1  0  0  0  0
  7  9  1  0  0  0  0
  8 12  1  0  0  0  0
  9 18  1  0  0  0  0
 10 13  2  0  0  0  0
 10 14  1  0  0  0  0
 11 12  2  0  0  0  0
 11 19  1  0  0  0  0
 12 20  1  0  0  0  0
 13 16  1  0  0  0  0
 13 21  1  0  0  0  0
 14 17  2  0  0  0  0
 14 22  1  0  0  0  0
 15 16  2  0  0  0  0
 15 17  1  0  0  0  0
 16 23  1  0  0  0  0
 17 24  1  0  0  0  0
M  END
",
    "mw": 259.34998199391646,
    "nbFragments": 1,
    "ocl": {
      "acceptorCount": 3,
      "coordinates": "!BkIbli~AexAjH@hc}bOvH@h\`BbOw~_x\`BbOvH?X\`BbOt",
      "donorCount": 1,
      "idCode": "fhypA@HHB@HdDIRjklkxIpSoTmKP@U@@@",
      "index": [
        0,
        179,
        198,
        233,
        129,
        0,
        17,
        42,
        128,
        0,
        0,
        16,
        0,
        0,
        128,
        64,
        4,
        1,
        160,
        128,
        0,
        2,
        4,
        0,
        20,
        0,
        36,
        1,
        12,
        8,
        8,
        136,
        32,
        16,
        16,
        8,
        36,
        0,
        34,
        0,
        0,
        1,
        20,
        0,
        128,
        12,
        5,
        8,
        2,
        0,
        144,
        128,
        0,
        0,
        4,
        0,
        0,
        0,
        64,
        0,
        130,
        64,
        16,
        0,
      ],
      "logP": 3.3200000412762165,
      "logS": -0.10999985039234161,
      "noStereoID": "fhypA@AZC\`HLDYIDdihdd]SfcN|uMHLAT@@_f@",
      "noStereoTautomerID": "fhypA@HHB@HdDIRjklkxIpSoTmKUUUP@Z|UcVCXmavOXSc}HXFmcxX~F@",
      "polarSurfaceArea": 105.34000015258789,
      "rotatableBondCount": 2,
      "stereoCenterCount": 0,
    },
    "unsaturation": 10,
  },
}
`;
